src/cpp/session/modules/automation/SessionAutomation.R

```diff
@@ -0,0 +1,343 @@
+      error = function(cnd) {
+         warning(conditionMessage(cnd))
+      },
+      
+      # Remove the callback when we're done.
+      finally = {
+        rm(list = as.character(id), envir = .rs.automation.callbacks)
+     }
+     
+   )
+   
+})
+
+.rs.addFunction("automation.onError", function(event)
+{
+   print(event)
+})
+
+.rs.addFunction("automation.onClose", function(event)
+{
+   print(event)
+})
+
+.rs.addFunction("automation.sendRequest", function(socket, method, params, callback = NULL)
+{
+   # Handle lazy callers.
+   if (is.function(params) && is.null(callback))
+   {
+      callback <- params
+      params <- list()
+   }
+   
+   # Generate an id for this request.
+   id <- .rs.automation.messageId
+   .rs.setVar("automation.messageId", .rs.automation.messageId + 1L)
+   
+   # Register a callback for this message id.
+   assign(as.character(id), callback, envir = .rs.automation.callbacks)
+   
+   # Generate the request.
+   request <- list(
+      id     = id,
+      method = method,
+      params = params
+   )
+   
+   # Attach a session ID if one is available.
+   if (!is.null(.rs.automation.sessionId))
+      request[["sessionId"]] <- .rs.automation.sessionId
+   
+   # Fire it off.
+   json <- .rs.toJSON(request, unbox = TRUE)
+   socket$send(json)
+   
+   # Return the id.
+   invisible(id)
+})
+
+.rs.addFunction("automation.sendSynchronousRequest", function(socket, method, params = list())
+{
+   # Drop NULL parameters.
+   params <- params[!vapply(params, is.null, FUN.VALUE = logical(1))]
+   
+   # Make sure parameters are named.
+   names(params) <- .rs.nullCoalesce(names(params), rep.int("", length(params)))
+   
+   # Send the request.
+   response <- NULL
+   callback <- function(response)
+   {
+      response <<- response
+   }
+   
+   .rs.automation.sendRequest(socket, method, params, callback)
+   
+   # Wait for a response. Pump the later event loop while we wait.
+   for (i in 1:100)
+   {
+      later::run_now()
+      if (!is.null(response))
+         break
+      
+      Sys.sleep(0.1)
+   }
+   
+   # Handle errors.
+   error <- response[["error"]]
+   if (!is.null(error)) {
+      fmt <- "execution of '%s' failed: %s [error code %i]"
+      msg <- sprintf(fmt, method, error[["message"]], error[["code"]])
+      stop(msg, call. = FALSE)
+   }
+   
+   # Return the received data.
+   response[["result"]]
+   
+})
+
+.rs.addFunction("automation.initialize", function(rstudioPath = NULL, port = 9999L)
+{
+   # Make sure all requisite packages are installed.
+   .rs.automation.installRequiredPackages()
+   
+   # Check for an existing session we can attach to.
+   baseUrl <- sprintf("http://localhost:%i", port)
+   jsonVersionUrl <- file.path(baseUrl, "json/version")
+   response <- .rs.tryCatch(httr::GET(jsonVersionUrl))
+   if (!inherits(response, "error"))
+      return(.rs.automation.attach(baseUrl))
+   
+   # No existing session; start a new one and attach to it.
```

Comment:
We should also support creating a Chrome / Chromium browser instance, and navigating to the RStudio Server URL.